File: intellex-backend/backend/app/api/routers/feeds.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from backend.app.api.deps import get_current_membership
from backend.app.api.schemas import (
    FeedSourceCreateRequest,
    FeedSourceListResponse,
    FeedSourceResponse,
    FeedSourceUpdateRequest,
)
from backend.app.collectors.rss import DEFAULT_FEEDS
from backend.app.db.models import OrganizationMemberModel
from backend.app.db.session import get_db
from backend.app.repositories.feed_source_repository import (
    FeedSourceRepository,
)

router = APIRouter(
    prefix="/feeds",
    tags=["Feeds"],
)


@router.get("/", response_model=FeedSourceListResponse)
async def get_feeds(
    db: Session = Depends(get_db),
    membership: OrganizationMemberModel = Depends(get_current_membership),
):
    repo = FeedSourceRepository(db)
    repo.seed_defaults_if_empty(membership.organization_id, DEFAULT_FEEDS)

    return FeedSourceListResponse(items=repo.list_all(membership.organization_id))


@router.post("/", response_model=FeedSourceResponse, status_code=201)
async def create_feed(
    payload: FeedSourceCreateRequest,
    db: Session = Depends(get_db),
    membership: OrganizationMemberModel = Depends(get_current_membership),
):
    repo = FeedSourceRepository(db)

    if repo.get_by_url(payload.url, membership.organization_id) is not None:
        raise HTTPException(
            status_code=409, detail="This feed URL is already configured"
        )

    try:
        return repo.create(payload.url, membership.organization_id, payload.label)
    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=409, detail="This feed URL is already configured"
        )


@router.patch("/{feed_id}", response_model=FeedSourceResponse)
async def update_feed(
    feed_id: str,
    payload: FeedSourceUpdateRequest,
    db: Session = Depends(get_db),
    membership: OrganizationMemberModel = Depends(get_current_membership),
):
    repo = FeedSourceRepository(db)

    model = repo.set_enabled(
        feed_id, membership.organization_id, payload.enabled
    )

    if model is None:
        raise HTTPException(status_code=404, detail="Feed not found")

    return model


@router.delete("/{feed_id}", status_code=204)
async def delete_feed(
    feed_id: str,
    db: Session = Depends(get_db),
    membership: OrganizationMemberModel = Depends(get_current_membership),
):
    repo = FeedSourceRepository(db)

    if not repo.delete(feed_id, membership.organization_id):
        raise HTTPException(status_code=404, detail="Feed not found")

File: intellex-frontend/frontend/src/lib/api/feeds.ts
import { apiClient } from "@/lib/api/client";
import type { FeedSource } from "@/lib/types";

interface RawFeedSource {
  id: string;
  url: string;
  label: string;
  enabled: boolean;
  created_at: string;
}

function toFeedSource(raw: RawFeedSource): FeedSource {
  return {
    id: raw.id,
    url: raw.url,
    label: raw.label,
    enabled: raw.enabled,
    createdAt: raw.created_at,
  };
}

export async function getFeeds(): Promise<FeedSource[]> {
  const { data } = await apiClient.get<{ items: RawFeedSource[] }>(
    "/feeds/"
  );

  return data.items.map(toFeedSource);
}

export async function createFeed(
  url: string,
  label = ""
): Promise<FeedSource> {
  const { data } = await apiClient.post<RawFeedSource>("/feeds/", {
    url,
    label,
  });

  return toFeedSource(data);
}

export async function deleteFeed(id: string): Promise<void> {
  await apiClient.delete(`/feeds/${id}`);
}

export async function setFeedEnabled(
  id: string,
  enabled: boolean
): Promise<FeedSource> {
  const { data } = await apiClient.patch<RawFeedSource>(`/feeds/${id}`, {
    enabled,
  });

  return toFeedSource(data);
}

export async function triggerIngestion(): Promise<{ status: string }> {
  const { data } = await apiClient.post<{ status: string }>(
    "/ingestion/trigger"
  );

  return data;
}
